### PR TITLE
Break some binding cycles

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/68110. In the process of going through all uses of `IsInterpolatedStringHandlerType` breaking things, I accidentally hit the same cycle, but in `IsComImport` (called by `IsExpressionOfComImportType`). I went through all the instances of these that I could find and constructed tests for these. The only location that I couldn't hit is https://github.com/dotnet/roslyn/blob/65fc4bd08b4bc15f6f00b529f47ffaeead03e434/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L6242
The test I added in SemanticErrorTests does not hit this. If anyone has suggestions for a test that could actually hit that, I'll add them and make the change there if necessary.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @AlekseyTs for review

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -7745,7 +7745,9 @@ private static void CombineExtensionMethodArguments(BoundExpression receiver, An
                 }
 
                 var overloadResolutionResult = OverloadResolutionResult<MethodSymbol>.GetInstance();
-                bool allowRefOmittedArguments = methodGroup.Receiver.IsExpressionOfComImportType();
+                // If we're in a parameter default value, this is an error scenario, so avoid checking for COM imported types to
+                // avoid a default value binding cycle.
+                bool allowRefOmittedArguments = !InParameterDefaultValue && methodGroup.Receiver.IsExpressionOfComImportType();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7750]: >!InParameterDefaultValue [](http://example.com/codeflow?start=48&length=24)

Why not also check for ```InAttributeArgument```?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -9075,7 +9077,9 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
         {
             Debug.Assert(receiver is not null);
             OverloadResolutionResult<PropertySymbol> overloadResolutionResult = OverloadResolutionResult<PropertySymbol>.GetInstance();
-            bool allowRefOmittedArguments = receiver.IsExpressionOfComImportType();
+            // We don't consider when we're in default parameter values to avoid cycles. This is an error scenario,
+            // so we don't care if we accidentally miss a parameter being applicable.
+            bool allowRefOmittedArguments = !InParameterDefaultValue && receiver.IsExpressionOfComImportType();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:9082]: >!InParameterDefaultValue [](http://example.com/codeflow?start=44&length=24)

Why not also check for ```InAttributeArgument```?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -9745,7 +9749,10 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
             else
             {
                 var result = OverloadResolutionResult<MethodSymbol>.GetInstance();
-                bool allowRefOmittedArguments = methodGroup.Receiver.IsExpressionOfComImportType();
+                // We check for default parameter value to avoid a cycle. If we're binding a parameter default value,
+                // this entire expression is bad, so we don't care whether this is a COM import type and we can safely
+                // just pass false here.
+                bool allowRefOmittedArguments = !this.InParameterDefaultValue && methodGroup.Receiver.IsExpressionOfComImportType();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:9755]: >!this.InParameterDefaultValue [](http://example.com/codeflow?start=48&length=29)

Why not also check for ```InAttributeArgument```?

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 4]
```diff
@@ -3830,7 +3830,10 @@ internal static bool AreRefsCompatibleForMethodConversion(RefKind x, RefKind y,
                     }
 
                     bool hasInterpolatedStringRefMismatch = false;
-                    if (argument is BoundUnconvertedInterpolatedString or BoundBinaryOperator { IsUnconvertedInterpolatedStringAddition: true }
+                    // We don't consider when we're in default parameter values to avoid cycles. This is an error scenario,
+                    // so we don't care if we accidentally miss a parameter being applicable.
+                    if (!_binder.InParameterDefaultValue
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3835]: >!_binder.InParameterDefaultValue [](http://example.com/codeflow?start=24&length=32)

Why not also check for ```InAttributeArgument```?

**Comment by AlekseyTs** [comment]
> The test I added in SemanticErrorTests does not hit this. If anyone has suggestions for a test that could actually hit that, I'll add them and make the change there if necessary.

I will be happy to debug the test with you offline in order to figure out how to hit the code path. Regardless, I think we should be proactive and add protection even if we aren't able to come up with scenario at the moment.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs [hunk 5]
```diff
@@ -129,6 +129,13 @@ public override Conversion GetMethodGroupFunctionPointerConversion(BoundMethodGr
 
         protected override Conversion GetInterpolatedStringConversion(BoundExpression source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
         {
+            if (_binder.InParameterDefaultValue || _binder.InAttributeArgument)
+            {
+                // We don't consider when we're in default parameter values or attributes to avoid cycles. This is an error scenario,
+                // so we don't care if we accidentally miss a parameter being applicable.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:135]: >miss a parameter being applicable [](http://example.com/codeflow?start=55&length=33)

Is this comment correct?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 6]
```diff
@@ -18172,5 +18173,75 @@ void M(string s)
...
+                #nullable enable
+
+                [A(3, X = 6)]
+                public struct A
+                {
+                    public int X;
+
+                    public A(int x, A a = M($"{1}")) { }
+
+                    public static void M(ref A a) {}
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:18225]: >void [](http://example.com/codeflow?start=34&length=4)

Minor point: Should the return type be `A`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 7]
```diff
@@ -24833,5 +24833,36 @@ public class C {
...
+            Assert.True(a.Parameters[1].HasExplicitDefaultValue);
+
+            compilation.VerifyDiagnostics(
+                // (3,2): error CS0653: Cannot apply attribute class 'A' because it is abstract
+                // [A(3, X = 6)]
+                Diagnostic(ErrorCode.ERR_AbstractAttributeClass, "A").WithArguments("A").WithLocation(3, 2),
+                // (8,32): error CS0144: Cannot create an instance of the abstract type or interface 'A'
+                //     public void M(int x, A a = new A()) { }
+                Diagnostic(ErrorCode.ERR_NoNewAbstract, "new A()").WithArguments("A").WithLocation(8, 32));
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:24866]: >} [](http://example.com/codeflow?start=8&length=1)

What leads to reentrancy in this case?

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 8]
```diff
@@ -24833,5 +24833,36 @@ public class C {
...
+            var source =
+                """
+                #nullable enable
+
+                [A(3, X = 6)]
+                public interface A
+                {
+                    public int X { get; set; }
+
+                    public void M(int x, A a = new A()) { }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:24849]: >public void M(int x, A a = new A()) { } [](http://example.com/codeflow?start=20&length=39)

Is this used? If not, consider removing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:18225]: It doesn't matter for the scenario.

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:24866]: It was an attempt to hit the re-entrancy in BindInterfaceCreationExpression, but as I mentioned, I wasn't able to get it to hit for that particular call site.

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs (hunk 8 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:24849]: Yes, I'm explicitly pulling on the default parameter value below.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:135]: Yes. In the cases checked, we're checking parameter applicability, either in an attribute constructor parameter or the default value of a parameter.

**Review by cston — APPROVED** [review]

**Pull request merged**